services/parse_pairs.go

```diff
@@ -0,0 +1,102 @@
+package services
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	. "github.com/beyondstorage/go-storage/v4/types"
+)
+
+var (
+	// ErrPairTypeNotParsable means the pair's type is not parseable.
+	ErrPairTypeNotParsable = NewErrorCode("type not parseable")
+	// ErrPairNotRegistered means the pair is not registered.
+	ErrPairNotRegistered = NewErrorCode("pair not registered")
+	// ErrPairValueInvalid means the pair's value is invalid.
+	ErrPairValueInvalid = NewErrorCode("pair value invalid")
+)
+
+// Parse a slice of Pairs from a map.
+func parseMap(ty string, m map[string]string) (ps []Pair, err error) {
+	servicePairMap, ok := servicePairMaps[ty]
+	if !ok {
+		err = fmt.Errorf("%w: %v", ErrServiceNotRegistered, ty)
+		return
+	}
+
+	for k, v := range m {
+		var pair Pair
+		if _, ok := globalPairMap[k]; ok {
+			pair, err = globalPairMap.parse(k, v)
+		} else {
+			pair, err = servicePairMap.parse(k, v)
+			pair.Key = ty + "_" + pair.Key
+		}
+		if err != nil {
+			return nil, err
+		}
+		ps = append(ps, pair)
+	}
+	return
+}
+
+func (m PairMap) parse(k string, v string) (pair Pair, err error) {
+	vType, ok := m[k]
+	if !ok {
+		err = fmt.Errorf("%w: %v", ErrPairNotRegistered, k)
+		return Pair{}, err
+	}
+
+	pair.Key = k
+
+	switch vType {
+	case "string":
+		pair.Value, err = v, nil
+	case "bool":
+		pair.Value, err = strconv.ParseBool(v)
+	case "int":
+		var i int64
+		i, err = strconv.ParseInt(v, 0, 0)
+		pair.Value = int(i)
+	case "int64":
+		pair.Value, err = strconv.ParseInt(v, 0, 64)
+	default:
+		return Pair{}, fmt.Errorf("%w: %v, %v", ErrPairTypeNotParsable, k, vType)
+	}
+
+	if err != nil {
+		pair = Pair{}
+		err = fmt.Errorf("%w: %v, %v, %v: %v", ErrPairValueInvalid, k, vType, v, err)
+	}
+	return
+}
+
+func parseListMode(s string) (ListMode, error) {
+	v64, err := strconv.ParseUint(s, 0, 8)
```

Comment:
the string is the human-readable representation, we should not allow user input list mode like `"2"` or `"0b11"`, just `"dir"` or `prefix|part`